src/main/java/edu/kit/compiler/optimizations/ArithmeticIdentitiesOptimization.java

```diff
@@ -0,0 +1,447 @@
+            assert constNode.getOpCode() == ir_opcode.iro_Const;
+
+            var restNode = getRestNode(nestedNode);
+            var leftConst = getConstOperand(nestedNode);
+            var rightConst = getConstValue(constNode);
+
+            if (restNode != null && leftConst != null && isAdditiveOperator(nestedNode)) {
+                // nestedNode is an additive operation with a Const predecessor
+                var accConst = accOperator.apply(leftConst, rightConst);
+                var newConst = graph.newConst(accConst);
+                var newNode = node.getOpCode() != nestedNode.getOpCode()
+                        ? graph.newSub(node.getBlock(), newConst, restNode)
+                        : graph.newAdd(node.getBlock(), restNode, newConst);
+
+                exchange(node, enqueued(newNode));
+            }
+        }
+
+        /**
+         * If possible, simplify the given Mul node using the associative
+         * property of multiplication. The right predecessor of the node
+         * must be a Const.
+         */
+        private void foldMultiplication(Mul node) {
+            assert node.getRight().getOpCode() == ir_opcode.iro_Const;
+
+            var restNode = getRestNode(node.getLeft());
+            var leftConst = getConstOperand(node.getLeft());
+            var rightConst = getConstValue(node.getRight());
+
+            if (restNode != null && leftConst != null
+                    && node.getLeft().getOpCode() == ir_opcode.iro_Mul) {
+                var constNode = graph.newConst(leftConst.mul(rightConst));
+                var newNode = graph.newMul(node.getBlock(), restNode, constNode);
+
+                exchange(node, enqueued(newNode));
+            }
+        }
+
+        /**
+         * Replace the given Div or Mode node. Result projections are replaced
+         * with `val`. Memory projections are replaced with `mem`.
+         */
+        private void exchangeDivOrMod(Node node, Node mem, Node val) {
+            assert node.getOpCode() == ir_opcode.iro_Div
+                    || node.getOpCode() == ir_opcode.iro_Mod;
+
+            for (var edge : BackEdges.getOuts(node)) {
+                if (edge.node.getMode().equals(Mode.getM())) {
+                    exchange(edge.node, mem);
+                } else if (edge.node.getMode().equals(val.getMode())) {
+                    exchange(edge.node, val);
+                } else {
+                    throw new UnsupportedOperationException(
+                            "Div control flow projections not supported");
+                }
+            }
+        }
+
+        /**
+         * Returns the first const predecessor of `node`.
+         */
+        private static TargetValue getConstOperand(Node node) {
+            for (var pred : node.getPreds()) {
+                if (pred.getOpCode() == ir_opcode.iro_Const) {
+                    return ((Const) pred).getTarval();
+                }
+            }
+            return null;
```

Comment:
As the method is private, this is rather a stylistic preference (feel free to ignore it, if you are of different opinion), but I would prefer either returning an `Optional` or, alternatively, throwing an exception (which would require checking at the call site that this case is impossible).

Reply:
I generally agree with you that returning `null` should be avoided. However, in this case of a private helper-method, I believe it's tolerable and allows the code to be a bit more concise.